### PR TITLE
RSDK-1870 Data Loss in ReadPCDtoBasicOctree

### PR DESCRIPTION
**Pull request by jeremyrhyde**
This PR fixes a bug with reading a PCD into a basic octree caused by the `bufio.Reader` being unable to be copied/duplicated. This led to the initial read to get metadata for the basic octree creation, leaving no bytes left for the actual adding process.

The fix has been to read the whole buffer to a string which then can be given to two separate `bufio.Reader` objects. 

**Testing:** Tested on multiple pcds of different sizes and types (ascii vs binary).

**Open Question:** In this PR the full PCD of the office is used for binary stress testing which takes approx 15 secs to be tested due to the presence of over 189,000 points. Do we want to use a small dataset for binary PCD testing or are we fine with the 15 sec time?

**Additional Details:**
asciiArtifactPath := "slam/rplidar_data/rplidar_data_0.pcd"
 - Size: 808
 - Time creation (pointcloud): 0.01 sec
 - Time creation (octree): 0.01 sec
 - Time testing: 0.12 sec
 
binaryArtifactPath := "slam/example_cartographer_outputs/viam-office-02-22-1/pointcloud/pointcloud_15.pcd"
 - Size: 1043
 - Time creation (pointcloud): 0.02 sec
 - Time creation (octree): 0.03 sec
 - Time testing: 0.11 sec

JIRA Ticket: [RSDK-1870](https://viam.atlassian.net/browse/RSDK-1870)

[RSDK-1870]: https://viam.atlassian.net/browse/RSDK-1870?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by nicksanford — CHANGES_REQUESTED** [review]

File: pointcloud/basic_octree_utils_test.go [hunk 1]
```diff
@@ -247,9 +247,11 @@ func validateMetadata(t *testing.T, bOct *BasicOctree) {
 	test.That(t, bOct.meta.MinY, test.ShouldEqual, metadata.MinY)
 	test.That(t, bOct.meta.MaxZ, test.ShouldEqual, metadata.MaxZ)
 	test.That(t, bOct.meta.MinZ, test.ShouldEqual, metadata.MinZ)
-	test.That(t, bOct.meta.TotalX(), test.ShouldAlmostEqual, metadata.TotalX())
-	test.That(t, bOct.meta.TotalY(), test.ShouldAlmostEqual, metadata.TotalY())
-	test.That(t, bOct.meta.TotalZ(), test.ShouldAlmostEqual, metadata.TotalZ())
+
+	tol := 0.0001
```
[the remark this answers is not in the document]
Comment (jeremyrhyde) [reply pointcloud/basic_octree_utils_test.go:251]: tolerance value

File: pointcloud/basic_octree_test.go [hunk 2]
```diff
@@ -557,8 +557,19 @@ func TestBasicOctreePointcloudIngestion(t *testing.T) {
 
 // Test the functionalities involved with converting a pcd into a basic octree.
 func TestReadBasicOctreeFromPCD(t *testing.T) {
-	artifactPath := "pointcloud/test_short.pcd"
+	t.Run("reading from binary PCD to octree", func(t *testing.T) {
+		binaryArtifactPath := "slam/example_cartographer_outputs/viam-office-02-22-1/pointcloud/pointcloud_15.pcd"
```
[the remark this answers is not in the document]
Comment (jeremyrhyde) [reply pointcloud/basic_octree_test.go:561]: In the additional details section, I note that the actual creation is only 0.3secs. I will look into on parallelizing the iterate calls

[the remark this answers is not in the document]
Comment (jeremyrhyde) [reply pointcloud/basic_octree_test.go:561]: Note: the time for octree creation includes adding all the points

[the remark this answers is not in the document]
Comment (tessavitabile) [reply pointcloud/basic_octree_test.go:561]: Can you find out what is taking 15s now? I want to make sure that Iterate() isn't slower than expected.

File: pointcloud/basic_octree_test.go [hunk 3]
```diff
@@ -557,8 +557,19 @@ func TestBasicOctreePointcloudIngestion(t *testing.T) {
 
 // Test the functionalities involved with converting a pcd into a basic octree.
 func TestReadBasicOctreeFromPCD(t *testing.T) {
```
Comment (tessavitabile) [inline pointcloud/basic_octree_test.go:559]: This test seems to check that all points in the pointcloud are also in the octree. Do you know why it passed previously? Does it fail now if the points are different?

**Review by tessavitabile — CHANGES_REQUESTED** [review]

File: pointcloud/basic_octree_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeremyrhyde) [reply pointcloud/basic_octree_test.go:561]: It is purely in the `validateBasicOctree` function. We do approx. 15x checks per node which in a balanced octree is `2*(numPoints) - 1`. In addition for each node we construct a new metadata object for comparison which involves iterating through all the points below it in the tree.

If this iterate that happens at every node is removed, the time goes down to 7 secs instead of 15 secs

[the remark this answers is not in the document]
Comment (jeremyrhyde) [reply pointcloud/basic_octree_test.go:561]: I could memoize the data here by starting at the leaves and going up to improve time but the check would still take at a minimum, 7 seconds. This function is only used for test evaluation so I can time box this change to be 2 hrs if this is desirable.

[the remark this answers is not in the document]
Comment (tessavitabile) [reply pointcloud/basic_octree_test.go:561]: Thanks, that makes sense. I think testing on a smaller octree is fine, and we don't need to optimize validateBasicOctree().

[the remark this answers is not in the document]
Comment (jeremyrhyde) [reply pointcloud/basic_octree_test.go:561]: @nicksanford based on the time sink being in the validate and caused not by the iterate, I don't think it's worth doing batching at present (the iterate takes approx. 0.2 sec on this large dataset and will be much smaller on a smaller dataset). Let me know if this works for you

[the remark this answers is not in the document]
Comment (jeremyrhyde) [reply pointcloud/basic_octree_test.go]: Test now takes 0.1 secs on the raw data from an rplidar (808 points)

File: pointcloud/basic_octree_utils_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (nicksanford) [reply pointcloud/basic_octree_utils_test.go]: Could you please also add a comment why this is needed?

File: pointcloud/basic_octree_test.go [hunk 4]
```diff
@@ -44,7 +45,9 @@ func checkPoints(t *testing.T, basicOct *BasicOctree, pointsAndData []PointAndDa
 // Helper function that makes and returns a PointCloud of a given type from an artifact path.
 func makeFullPointCloudFromArtifact(t *testing.T, artifactPath string, pcType PCType) (PointCloud, error) {
 	t.Helper()
-	pcdFile, err := os.Open(artifact.MustPath(artifactPath))
+
+	path := filepath.Clean(artifact.MustPath(artifactPath))
+	pcdFile, err := os.Open(path)
```
Comment (nicksanford) [inline pointcloud/basic_octree_test.go:50]: Don't we need to defer close?

File: pointcloud/basic_octree_utils_test.go [hunk 5]
```diff
@@ -247,9 +247,11 @@ func validateMetadata(t *testing.T, bOct *BasicOctree) {
 	test.That(t, bOct.meta.MinY, test.ShouldEqual, metadata.MinY)
 	test.That(t, bOct.meta.MaxZ, test.ShouldEqual, metadata.MaxZ)
 	test.That(t, bOct.meta.MinZ, test.ShouldEqual, metadata.MinZ)
-	test.That(t, bOct.meta.TotalX(), test.ShouldAlmostEqual, metadata.TotalX())
-	test.That(t, bOct.meta.TotalY(), test.ShouldAlmostEqual, metadata.TotalY())
-	test.That(t, bOct.meta.TotalZ(), test.ShouldAlmostEqual, metadata.TotalZ())
+
+	tolerance := 0.0001
```
Comment (nicksanford) [inline pointcloud/basic_octree_utils_test.go:251]: Can we please add a comment why this is needed?

File: .artifact/tree.json [hunk 6]
```diff
@@ -50623,6 +50623,7 @@
       "size": 145250924
     },
     "example_cartographer_outputs": {
+      "position": {},
```
Comment (nicksanford) [inline .artifact/tree.json:50626]: Why is this here?

File: pointcloud/basic_octree_utils_test.go (hunk 5 above)
Comment (jeremyrhyde) [reply pointcloud/basic_octree_utils_test.go:251]: done

File: pointcloud/basic_octree_test.go (hunk 4 above)
Comment (jeremyrhyde) [reply pointcloud/basic_octree_test.go:50]: added

File: pointcloud/basic_octree_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeremyrhyde) [reply pointcloud/basic_octree_test.go]:
```
carto output: 180,000 points. + 15 secs
pointcloud/bun090.pcd: 30,000 points + 2.7 sec
rplidar_data: 800 points + 0.1 sec
```

@tessavitabile @nicksanford please confirm that we are alright with the 800 point pcd file instead of the 30,000 that takes 2.7 sec

[the remark this answers is not in the document]
Comment (tessavitabile) [reply pointcloud/basic_octree_test.go]: yes

File: .artifact/tree.json (hunk 6 above)
Comment (jeremyrhyde) [reply .artifact/tree.json]: not sure why, and removed

File: pointcloud/basic_octree_test.go (hunk 3 above)
Comment (jeremyrhyde) [reply pointcloud/basic_octree_test.go:564]: See conversation here: https://viaminc.slack.com/archives/C036N5GJFLM/p1678115565635339

File: pointcloud/basic_octree_utils_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeremyrhyde) [reply pointcloud/basic_octree_utils_test.go]: done

File: pointcloud/basic_octree_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeremyrhyde) [reply pointcloud/basic_octree_test.go]: done

File: go.mod [hunk 7]
```diff
@@ -64,8 +64,8 @@ require (
 	github.com/nfnt/resize v0.0.0-20180221191011-83c6a9932646
 	github.com/pion/mediadevices v0.4.0
 	github.com/pion/rtp v1.7.13
-	github.com/pixiv/go-libjpeg v0.0.0-20190822045933-3da21a74767d
 	github.com/pion/webrtc/v3 v3.1.54
+	github.com/pixiv/go-libjpeg v0.0.0-20190822045933-3da21a74767d
```
Comment (jeremyrhyde) [inline go.mod:68]: `make build lint`

File: pointcloud/basic_octree_test.go (hunk 3 above)
Comment (tessavitabile) [reply pointcloud/basic_octree_test.go:564]: Just to confirm, without the fix, the test does fail on large input sizes? And we're now testing large enough input sizes that we would catch the regression?

**Review by tessavitabile — APPROVED** [review]
LGTM mod question

**Review by jeremyrhyde — COMMENTED** [review]
@tessavitabile This type of error can only occur on files with ~247 points. Both of our files have +800 so we should detect it if there is an issue

**Review by nicksanford — APPROVED** [review]
LGTM! Thanks for root causing! Great find!!!
